### PR TITLE
[orchestration] resort to upstart and systemd for agent orchestration

### PR DESCRIPTION
**Pull request by truthbk**
### What does this PR do?
Agents are pretty much decoupled, process orchestration can be deferred to upstart or systemd. PID files were introduced to all agents to allow process tracking. Additionally agents can shutdown sanely if disabled, and also hard-disabled by resorting to `systemctl`, `initctl`. Rate limiting has been added to the systemd/upstart such that if a process has to be restarted 5 times in a 10 second window it'll cease to reattempt.

### Motivation
The agent should not be a process orchestrator.

### Additional Notes

Currently `datadog-agent` is the infra group, but other agents depend on it and are started after it has come up. That said, a restart to `datadog-agent` will restart all agents. We can easily change this behavior if deemed desirable.

### CODE REVIEW

**Comment by xvello** [comment]
It does make sense, although I have a question and a request:

- is agent6 still going to manage jmxfetch? If so why? IIRC in agent5 jmxfetch was started by supervisord
- this will require work on the docker image, to select and configure an init system. Can we target `beta10` for this, and release `beta9` with the current logic?

**Comment by truthbk** [comment]
@xvello currently we decided to keep jmx as a long-running check, I believe the plan is to keep it that way, at least for now.

**Comment by olivielpeau** [comment]
Haven't tested this (yet), but it looks 👌 👌 👌 

Blockers/TODO/things to confirm for Croissant (that I've noticed):
- [ ] check that the options that are currently passed as env vars to the process agent can be picked up by the process agent on its own: https://github.com/DataDog/datadog-agent/blob/6.0.0-beta.8/pkg/collector/corechecks/embed/process_agent.go#L79-L85 (they probably are, let's double-check)
- [ ] same thing for the trace-agent: https://github.com/DataDog/datadog-agent/blob/6.0.0-beta.8/pkg/collector/corechecks/embed/apm.go#L69-L72 (not so sure about this one, especially for the hostname)
- [ ] has anything changed to enable/disable the process agent? If so we should document it (and update our config mgmt plugins)
- [ ] has anything changed to enable/disable the trace agent? If so we should document it (and update our config mgmt plugins)

Just to clarify the behavior of JMXFetch in Agent5: supervisord would start a python script (`jmfetch.py`) that would in turn start jmxfetch as a subprocess (passing the right configuration options).
Right now, to declare jmxfetch purely as a service in agent6, I think we'd need to have a similar wrapper (to pass the right options to the jmxfetch jar), which would mean quite some additional work.

**Comment by xvello** [comment]
Makes sense for JMX 👍 

I'm gonna start work on an image using [s6 init](https://github.com/just-containers/s6-overlay), starting the three processes independently like supervisord does in the agent5 images.

My concern is the trace-agent hostname, here's the sniplet from agent5 entrypoint:
```bash
if [ -z $DD_HOSTNAME ] && [ $DD_APM_ENABLED ]; then
  # When starting up the trace-agent without an explicit hostname
  # we need to ensure that the trace-agent will report as the same host as the
  # infrastructure agent.
  # To do this, we execute some of dd-agent's python code and expose the hostname
  # as an env var
  export DD_HOSTNAME=`python -c "from utils.hostname import get_hostname; print get_hostname()"`
fi
```

I see several options here:
  * get `trace-agent` to import `pkg/hostname` and be autononous
  * introduce a standalone `agent hostname` command outputing the hostname from the same logic it uses, even if not started, and inject that as an envvar to trace-agent. Integrating that in s6 would be not too hard, dunno about upstart
  * get `trace-agent` get the hostname from `agent status`, that'd make infra-agent a startup dependency of trace-agent

The first solution would be the best mid-term solution, as it would allow trace-agent to be completely autonomous for use in kubernetes/cloudformation sidecars (just needing a UDP dsd available, but no startup dependency)

**Comment by xvello** [comment]
It be able to run the three processes in 3 separate containers pertaining to the same pod, we'd need to shared volumes. The obvious one is `/var/log/datadog/` (so the infra agent flare can get them).
We'll also need to share the pid files and the API auth token. Shall they go to a shared `..../run/` folder?

The requirement for shared volumes is that they start empty.

File: omnibus/config/templates/datadog-agent/upstart_debian.trace.conf.erb [hunk 1]
```diff
@@ -15,7 +15,7 @@ env DD_LOG_TO_CONSOLE=false
 setuid dd-agent
 
 script
-  exec <%= install_dir %>/embedded/bin/trace-agent --config <%= etc_dir %>/trace_agent.conf --pid <%= install_dir %>/run/trace-agent.pid
+  exec <%= install_dir %>/embedded/bin/trace-agent --config <%= etc_dir %>/datadog.yaml --pid <%= install_dir %>/run/trace-agent.pid
```
Comment (xvello) [inline omnibus/config/templates/datadog-agent/upstart_debian.trace.conf.erb]: does it need to be `--ddconfig` too?

Comment (nmuesch) [reply omnibus/config/templates/datadog-agent/upstart_debian.trace.conf.erb]: Yupp, missed a save. Thanks 🙇

**Review by olivielpeau — DISMISSED** [review]
Looks great! 👏 

A couple of general comments, can/should be addressed in a different PR:

* we should remove the old INI trace-agent.conf.example file from all packages.
* we should remove the apm and process check example/default yamls from the linux and windows packages (except osx basically)

**Pull request merged**